share/man/man4/rnd.4
.\"	$NetBSD: rnd.4,v 1.5 1999/03/16 01:19:17 garbled Exp $
.\"
.\" Copyright (c) 1997 Michael Graff
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
.\" BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
.\" AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd October 12, 1997
.Dt RND 4
.Os
.Sh NAME
.Nm rnd
.Nd in kernel entropy collection and random number generation
.Sh SYNOPSIS
.Cd pseudo-device rnd
.Sh DESCRIPTION
The
.Nm
pseudo-device uses event timing information collected from many
devices, and mixes this into an entropy pool.  This pool is stirred
with a cryptographically strong hash function when data is extracted
from the pool.
.Sh INTERNAL ENTROPY POOL MANAGEMENT
When a hardware event occurs (such as completion of a hard drive
transfer or an interrupt from a network device) a timestamp is
generated.  This timestamp is compared to the previous timestamp
recorded for the device, and the first, second, and third order
differentials are calculated.
.Pp
If any of these differentials is zero, no entropy is assumed to
have been gathered.  If all are non-zero, one bit is assumed.
Next, data is mixed into the entropy pool using an LFSR (linear
feedback shift register).
.Pp
To extract data from the entropy pool, a cryptographically strong hash
function is used.  The output of this hash is mixed back into the pool
using the LFSR, and then folded in half before being returned to the
caller.
.Pp
Mixing the actual hash into the pool causes the next extraction to
return a different value, even if no timing events were added to the
pool.  Folding the data in half prevents the caller to derive the
actual hash of the pool, preventing some attacks.
.Sh USER ACCESS
User code can obtain random values from the kernel in two ways.
.Pp
Reading from
.Pa /dev/random
will only return values while sufficient entropy exists in the
internal pool.  When sufficent entropy does not exist, EAGAIN is
returned for non-blocking reads, or the read will block for blocking
reads.
.Pp
Reading from
.Pa /dev/urandom
will return as many values as requested, even when the entropy pool is
empty.  This data is not as good as reading from
.Pa /dev/random
since when the pool is empty, data is still returned, degenerating to a
pseudo-random generator.
.Pp
Writing to either device will mix the data written into the pool using
the LFSR as above, without modifying the entropy estimation for the
pool.
.Sh RANDOM SOURCE STRUCTURE
Each source has a state structure which the kernel uses to hold the
timing information and other state for that source.
.Bd -literal -offset indent
typedef struct {
        char            name[16];
        u_int32_t       last_time;
        u_int32_t       last_delta;
        u_int32_t       last_delta2;
        u_int32_t       total;
        u_int32_t       type;
	u_int32_t	flags;
} rndsource_t;
.Ed
.Pp
This structure holds the internal representation of a device's timing
state.  The
.Va name
field holes the device name, as known to the kernel.  The
.Va last_time
entry is the timestamp of the last time this device generated an
event.  It is for internal use only, and not in any specific
representation.  The
.Va last_delta
and
.Va last_delta2
fields hold the last first- and second-order deltas.  The
.Va total
field holds a count of how many bits this device has potentially
generated.  This is not the same as how many bits were used from it.
The
.Va type
field holds the device type.
.Pp
.Bl -tag -width RND_TYPE_DISK
Currently, these types are defined:
.It Dv RND_TYPE_DISK
The device is a physical hard drive.
.It Dv RND_TYPE_NET
The device is a network interface.  By default, timing information is
collected from this source type, but entropy is not estimated.
.It Dv RND_TYPE_TAPE
The device is a tape device.
.It Dv RND_TYPE_TTY
The device is a terminal, mouse, or other user input device.
.El
.Pp
.Va flags
is a bitfield.
.Bl -tag -width RND_FLAG_NO_ESTIMATE
.It Dv RND_FLAG_NO_ESTIMATE
Do not assume any entropy is in the timing information.
.It Dv RND_FLAG_NO_COLLECT
Do not even add timing information to the pool.
.El
.Sh IOCTL
Various
.Xr ioctl 2
functions are available to control device behavior, gather statistics,
and add data to the entropy pool.  These are all defined in the
.Aq Pa sys/rnd.h
file, along with the data types and constants.
.Pp
.Bl -tag -width RNDADDTOENTCNT
.It Dv RNDGETENTCNT
.Pq Li "u_int32_t"
Return the current entropy count (in bits).
.It Dv RNDGETSRCNUM
.Pq Li "rndstat_t"
.Bd -literal -offset indent
typedef struct {
        u_int32_t       start;
        u_int32_t       count;
        rndsource_t     source[RND_MAXSTATCOUNT];
} rndstat_t;
.Ed
.Pp
Return data for sources, starting at
.Va start
and returning at most
.Va count
sources.
.Pp
The values returned are actual in-kernel snapshots of the entropy
status for devices.  Leaking the internal timing information will
weaken security.
.It Dv RNDGETSRCNAME
.Pq Li "rndstat_name_t"
.Bd -literal -offset indent
typedef struct {
        char            name[16];
        rndsource_t     source;
} rndstat_name_t;
.Ed
.Pp
Return the device state for a named device.
.It Dv RNDCTL
.Pq Li "rndctl_t"
.Bd -literal -offset indent
typedef struct {
        char            name[16];
        u_int32_t       type;
        u_int32_t       flags;
        u_int32_t       mask;
} rndctl_t;
.Ed
.Pp
Change bits in the device state information.  If
.Va type
is 0xff, only the device name stored in
.Va name
is used.  If it is any other value, all devices of type
.Va type
are altered.  This allows all network interfaces to be disabled for
entropy collection with one call, for example.
The
.Va flags
and
.Va mask
work together to change flag bits.  The
.Va mask
field specifies which bits in
.Va flags
are to be set or cleared.
.It Dv RNDADDDATA
.Pq Li "rnddata_t"
.Bd -literal -offset indent
typedef struct {
        u_int32_t       len;
        u_int32_t       entropy;
        u_char          data[RND_POOLWORDS * 4];
} rnddata_t;
.Ed
.El
.Sh FILES
.Bl -tag -width /dev/urandomx -compact
.It Pa /dev/random
Returns ``good'' values only
.It Pa /dev/urandom
Always returns data, degenerates to a pseudo-random generator
.El
.Sh HISTORY
The random device was first made available in
.Nx 1.3 .
.Sh AUTHOR
This implementation was written by Michael Graff <explorer@flame.org>
using ideas and algorithms gathered from many sources, including
the driver written by Ted Ts'o.
.Sh SEE ALSO
.Xr rndctl 8 ,
.Xr rnd 9 .


sys/dev/rnd.c
/*	$NetBSD: rnd.c,v 1.16 1999/06/12 10:58:47 pk Exp $	*/

/*-
 * Copyright (c) 1997 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Michael Graff <explorer@flame.org>.  This code uses ideas and
 * algorithms from the Linux driver written by Ted Ts'o.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *        This product includes software developed by the NetBSD
 *        Foundation, Inc. and its contributors.
 * 4. Neither the name of The NetBSD Foundation nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/param.h>
#include <sys/ioctl.h>
#include <sys/fcntl.h>
#include <sys/select.h>
#include <sys/poll.h>
#include <sys/malloc.h>
#include <sys/proc.h>
#include <sys/kernel.h>
#include <sys/conf.h>
#include <sys/systm.h>
#include <sys/rnd.h>
#include <sys/vnode.h>
#include <sys/pool.h>

#ifdef RND_DEBUG
#define DPRINTF(l,x)      if (rnd_debug & (l)) printf x
int     rnd_debug = 0;
#else
#define DPRINTF(l,x)
#endif

#define RND_DEBUG_WRITE		0x0001
#define RND_DEBUG_READ		0x0002
#define RND_DEBUG_IOCTL		0x0004
#define RND_DEBUG_SNOOZE	0x0008

/*
 * list devices attached
 */
/* #define RND_VERBOSE */

/*
 * Use the extraction time as a somewhat-random source
 */
#ifndef RND_USE_EXTRACT_TIME
#define RND_USE_EXTRACT_TIME 1
#endif

/*
 * The size of a temporary buffer, malloc()ed when needed, and used for
 * reading and writing data.
 */
#define RND_TEMP_BUFFER_SIZE	128

/*
 * This is a little bit of state information attached to each device that we
 * collect entropy from.  This is simply a collection buffer, and when it
 * is full it will be "detached" from the source and added to the entropy
 * pool after entropy is distilled as much as possible.
 */
#define RND_SAMPLE_COUNT	64	/* collect N samples, then compress */
typedef struct _rnd_sample_t {
	SIMPLEQ_ENTRY(_rnd_sample_t)	next;
	rndsource_t    *source;
	int		cursor;
	int		entropy;
	u_int32_t	ts[RND_SAMPLE_COUNT];
	u_int32_t	values[RND_SAMPLE_COUNT];
} rnd_sample_t;

/*
 * the event queue.  Fields are altered at an interrupt level.
 */
volatile int			rnd_timeout_pending;
SIMPLEQ_HEAD(, _rnd_sample_t)	rnd_samples;

/*
 * our select/poll queue
 */
struct selinfo rnd_selq;

/*
 * Set when there are readers blocking on data from us
 */
#define RND_READWAITING 0x00000001
volatile u_int32_t  rnd_status;

/*
 * Memory pool.
 */
struct pool rnd_mempool;

/*
 * our random pool.  This is defined here rather than using the general
 * purpose one defined in rndpool.c
 */
rndpool_t   rnd_pool;

/*
 * This source is used to easily "remove" queue entries when the source
 * which actually generated the events is going away.
 */
static rndsource_t rnd_source_no_collect = {
	{ 'N', 'o', 'C', 'o', 'l', 'l', 'e', 'c', 't', 0, 0, 0, 0, 0, 0, 0 },
	0, 0, 0, 0,
	(RND_FLAG_NO_COLLECT | RND_FLAG_NO_ESTIMATE | RND_TYPE_UNKNOWN),
	NULL
};

void	rndattach __P((int));
int	rndopen __P((dev_t, int, int, struct proc *));
int	rndclose __P((dev_t, int, int, struct proc *));
int	rndread __P((dev_t, struct uio *, int));
int	rndwrite __P((dev_t, struct uio *, int));
int	rndioctl __P((dev_t, u_long, caddr_t, int, struct proc *));
int	rndpoll __P((dev_t, int, struct proc *));

static inline void	rnd_wakeup_readers(void);
static inline u_int32_t rnd_estimate_entropy(rndsource_t *, u_int32_t);
static inline u_int32_t rnd_timestamp(void);
static 	      void	rnd_timeout(void *);

static int		rnd_ready = 0;
static int		rnd_have_entropy = 0;

LIST_HEAD(, __rndsource_element)	rnd_sources;

/*
 * Generate a 32-bit timestamp.  This should be more machine dependant,
 * using cycle counters and the like when possible.
 */
static inline u_int32_t
rnd_timestamp()
{
	struct timeval	tv;
	u_int32_t	t;

	microtime(&tv);

	t = tv.tv_sec * 1000000 + tv.tv_usec;

	return t;
}

/*
 * Check to see if there are readers waiting on us.  If so, kick them.
 */
static inline void
rnd_wakeup_readers()
{
	/*
	 * If we have added new bits, and now have enough to do something,
	 * wake up sleeping readers.
	 */
	if (rndpool_get_entropy_count(&rnd_pool) > RND_ENTROPY_THRESHOLD * 8) {
		if (rnd_status & RND_READWAITING) {
			DPRINTF(RND_DEBUG_SNOOZE,
				("waking up pending readers.\n"));
			rnd_status &= ~RND_READWAITING;
			wakeup(&rnd_selq);
		}
		selwakeup(&rnd_selq);

		/*
		 * Allow open of /dev/random now, too.
		 */
		rnd_have_entropy = 1;
	}
}

/*
 * Use the timing of the event to estimate the entropy gathered.
 * If all the differentials (first, second, and third) are non-zero, return
 * non-zero.  If any of these are zero, return zero.
 */
static inline u_int32_t
rnd_estimate_entropy(rs, t)
	rndsource_t *rs;
	u_int32_t t;
{
	int32_t		delta;
	int32_t		delta2;
	int32_t		delta3;

	/*
	 * If the time counter has overflowed, calculate the real difference.
	 * If it has not, it is simplier.
	 */
	if (t < rs->last_time)
		delta = UINT_MAX - rs->last_time + t;
	else
		delta = rs->last_time - t;

	if (delta < 0)
		delta = -delta;

	/*
	 * Calculate the second and third order differentials
	 */
	delta2 = rs->last_delta - delta;
	if (delta2 < 0)
		delta2 = -delta2;

	delta3 = rs->last_delta2 - delta2;
	if (delta3 < 0)
		delta3 = -delta3;

	rs->last_time = t;
	rs->last_delta = delta;
	rs->last_delta2 = delta2;

	/*
	 * If any delta is 0, we got no entropy.  If all are non-zero, we
	 * might have something.
	 */
	if (delta == 0 || delta2 == 0 || delta3 == 0)
		return 0;

	return 1;
}

/*
 * attach the random device, and initialize the global random pool
 * for our use.
 */
void
rndattach(num)
	int num;
{

	rnd_init();
}

void
rnd_init(void)
{

	if (rnd_ready)
		return;

	LIST_INIT(&rnd_sources);
	SIMPLEQ_INIT(&rnd_samples);

	pool_init(&rnd_mempool, sizeof(rnd_sample_t), 0, 0, 0, "rndsample",
		  0, NULL, NULL, NULL);

	rndpool_init(&rnd_pool);

	rnd_ready = 1;

#ifdef RND_VERBOSE
	printf("Random device ready\n");
#endif
}

int
rndopen(dev, flags, ifmt, p)
	dev_t dev;
	int flags, ifmt;
	struct proc *p;
{

	if (rnd_ready == 0)
		return (ENXIO);

	if (minor(dev) == RND_DEV_URANDOM)
		return (0);

	/*
	 * If this is the strong random device and we have never collected
	 * entropy (or have not yet) don't allow it to be opened.  This will
	 * prevent waiting forever for something that just will not appear.
	 */
	if (minor(dev) == RND_DEV_RANDOM) {
		if (rnd_have_entropy == 0)
			return (ENXIO);
		else
			return (0);
	}

	return (ENXIO);
}

int
rndclose(dev, flags, ifmt, p)
	dev_t dev;
	int flags, ifmt;
	struct proc *p;
{

	return (0);
}

int
rndread(dev, uio, ioflag)
	dev_t dev;
	struct uio *uio;
	int ioflag;
{
	int		ret;
	u_int32_t       nread;
	int		n;
	int		s;
	u_int8_t	*buf;
	u_int32_t	mode;
	u_int32_t	entcnt;

	DPRINTF(RND_DEBUG_READ,
		("Random:  Read of %d requested, flags 0x%08x\n",
		 uio->uio_resid, ioflag));

	if (uio->uio_resid == 0)
		return (0);

	switch (minor(dev)) {
	case RND_DEV_RANDOM:
		mode = RND_EXTRACT_GOOD;
		break;
	case RND_DEV_URANDOM:
		mode = RND_EXTRACT_ANY;
		break;
	default:
		/* Can't happen, but this is cheap */
		return (ENXIO);
	}

	ret = 0;

	buf = malloc(RND_TEMP_BUFFER_SIZE, M_TEMP, M_WAITOK);

	while (uio->uio_resid > 0) {
		n = min(RND_TEMP_BUFFER_SIZE, uio->uio_resid);

		/*
		 * Make certain there is data available.  If there
		 * is, do the I/O even if it is partial.  If not,
		 * sleep unless the user has requested non-blocking
		 * I/O.
		 */
		for (;;) {
			/*
			 * If not requesting strong randomness, we
			 * can always read.
			 */
			if (mode == RND_EXTRACT_ANY)
				break;

			/*
			 * How much entropy do we have?  If it is enough for
			 * one hash, we can read.
			 */
			s = splsoftclock();
			entcnt = rndpool_get_entropy_count(&rnd_pool);
			splx(s);
			if (entcnt >= RND_ENTROPY_THRESHOLD * 8)
				break;

			/*
			 * Data is not available.
			 */
			if (ioflag & IO_NDELAY) {
				ret = EWOULDBLOCK;
				goto out;
			}

			rnd_status |= RND_READWAITING;
			ret = tsleep(&rnd_selq, PRIBIO|PCATCH,
				     "rndread", 0);

			if (ret)
				goto out;
		}

		nread = rnd_extract_data(buf, n, mode);

		/*
		 * copy (possibly partial) data to the user.
		 * If an error occurs, or this is a partial
		 * read, bail out.
		 */
		ret = uiomove((caddr_t)buf, nread, uio);
		if (ret != 0 || nread != n)
			goto out;
	}

out:
	free(buf, M_TEMP);
	return (ret);
}

int
rndwrite(dev, uio, ioflag)
	dev_t dev;
	struct uio *uio;
	int ioflag;
{
	u_int8_t	*buf;
	int		ret;
	int		n;
	int		s;

	DPRINTF(RND_DEBUG_WRITE,
		("Random:  Write of %d requested\n", uio->uio_resid));

	if (uio->uio_resid == 0)
		return (0);

	ret = 0;

	buf = malloc(RND_TEMP_BUFFER_SIZE, M_TEMP, M_WAITOK);

	while (uio->uio_resid > 0) {
		n = min(RND_TEMP_BUFFER_SIZE, uio->uio_resid);

		ret = uiomove((caddr_t)buf, n, uio);
		if (ret != 0)
			break;

		/*
		 * Mix in the bytes.
		 */
		s = splsoftclock();
		rndpool_add_data(&rnd_pool, buf, n, 0);
		splx(s);

		DPRINTF(RND_DEBUG_WRITE, ("Random:  Copied in %d bytes\n", n));
	}

	free(buf, M_TEMP);
	return (ret);
}

int
rndioctl(dev, cmd, addr, flag, p)
	dev_t dev;
	u_long cmd;
	caddr_t addr;
	int flag;
	struct proc *p;
{
	int			 ret;
	rndsource_element_t	*rse;
	rndstat_t		*rst;
	rndstat_name_t		*rstnm;
	rndctl_t		*rctl;
	rnddata_t		*rnddata;
	u_int32_t		 count;
	u_int32_t		 start;
	int			 s;

	ret = 0;

	switch (cmd) {

	  /*
	   * Handled in upper layer really, but we have to return zero
	   * for it to be accepted by the upper layer.
	   */
	case FIONBIO:
	case FIOASYNC:
		break;

	case RNDGETENTCNT:
		s = splsoftclock();
		*(u_int32_t *)addr = rndpool_get_entropy_count(&rnd_pool);
		splx(s);

		break;

	case RNDGETSRCNUM:
		if ((ret = suser(p->p_ucred, &p->p_acflag)) != 0)
			return (ret);

		rst = (rndstat_t *)addr;

		if (rst->count == 0)
			break;

		if (rst->count > RND_MAXSTATCOUNT)
			return (EINVAL);

		/*
		 * find the starting source by running through the
		 * list of sources.
		 */
		rse = rnd_sources.lh_first;
		start = rst->start;
		while (rse != NULL && start >= 1) {
			rse = rse->list.le_next;
			start--;
		}

		/*
		 * Return up to as many structures as the user asked
		 * for.  If we run out of sources, a count of zero
		 * will be returned, without an error.
		 */
		for (count = 0 ; count < rst->count && rse != NULL ; count++) {
			bcopy(&rse->data, &rst->source[count],
			      sizeof(rndsource_t));
			rse = rse->list.le_next;
		}

		rst->count = count;

		break;

	case RNDGETSRCNAME:
		if ((ret = suser(p->p_ucred, &p->p_acflag)) != 0)
			return (ret);

		/*
		 * scan through the list, trying to find the name
		 */
		rstnm = (rndstat_name_t *)addr;
		rse = rnd_sources.lh_first;
		while (rse != NULL) {
			if (strncmp(rse->data.name, rstnm->name, 16) == 0) {
				bcopy(&rse->data, &rstnm->source,
				      sizeof(rndsource_t));

				return 0;
			}
			rse = rse->list.le_next;
		}

		ret = ENOENT;  /* name not found */

		break;

	case RNDCTL:
		if ((ret = suser(p->p_ucred, &p->p_acflag)) != 0)
			return (ret);

		/*
		 * set flags to enable/disable entropy counting and/or
		 * collection
		 */
		rctl = (rndctl_t *)addr;
		rse = rnd_sources.lh_first;

		/*
		 * flags set apply to all sources of this type
		 */
		if (rctl->type != 0xff) {
			while (rse != NULL) {
				if (rse->data.type == rctl->type) {
					rse->data.flags &= ~rctl->mask;
					rse->data.flags |= (rctl->flags
							   & rctl->mask);
				}
				rse = rse->list.le_next;
			}

			return 0;
		}

		/*
		 * scan through the list, trying to find the name
		 */
		while (rse != NULL) {
			if (strncmp(rse->data.name, rctl->name, 16) == 0) {
				rse->data.flags &= ~rctl->mask;
				rse->data.flags |= (rctl->flags & rctl->mask);

				return 0;
			}
			rse = rse->list.le_next;
		}

		ret = ENOENT;  /* name not found */

		break;

	case RNDADDDATA:
		if ((ret = suser(p->p_ucred, &p->p_acflag)) != 0)
			return (ret);

		rnddata = (rnddata_t *)addr;

		s = splsoftclock();
		rndpool_add_data(&rnd_pool, rnddata->data, rnddata->len,
				 rnddata->entropy);

		rnd_wakeup_readers();
		splx(s);

		break;

	default:
		return (EINVAL);
	}

	return (ret);
}

int
rndpoll(dev, events, p)
	dev_t dev;
	int events;
	struct proc *p;
{
	int revents;
	int s;
	u_int32_t entcnt;

	/*
	 * we are always writable
	 */
	revents = events & (POLLOUT | POLLWRNORM);

	/*
	 * Save some work if not checking for reads
	 */
	if ((events & (POLLIN | POLLRDNORM)) == 0)
		return (revents);

	/*
	 * If the minor device is not /dev/random, we are always readable.
	 */
	if (minor(dev) != RND_DEV_RANDOM) {
		revents |= events & (POLLIN | POLLRDNORM);
		return (revents);
	}

	/*
	 * make certain we have enough entropy to be readable
	 */
	s = splsoftclock();
	entcnt = rndpool_get_entropy_count(&rnd_pool);
	splx(s);

	if (entcnt >= RND_ENTROPY_THRESHOLD * 8)
		revents |= events & (POLLIN | POLLRDNORM);
	else
		selrecord(p, &rnd_selq);

	return (revents);
}

static rnd_sample_t *
rnd_sample_allocate(rndsource_t *source)
{
	rnd_sample_t *c;
	int s;

	s = splhigh();
	c = pool_get(&rnd_mempool, PR_WAITOK);
	splx(s);
	if (c == NULL)
		return (NULL);

	c->source = source;
	c->cursor = 0;
	c->entropy = 0;

	return (c);
}

/*
 * don't wait on allocation.  to be used in an interrupt context.
 */
static rnd_sample_t *
rnd_sample_allocate_isr(rndsource_t *source)
{
	rnd_sample_t *c;
	int s;

	s = splhigh();
	c = pool_get(&rnd_mempool, 0);
	splx(s);
	if (c == NULL)
		return (NULL);

	c->source = source;
	c->cursor = 0;
	c->entropy = 0;

	return (c);
}

static void
rnd_sample_free(rnd_sample_t *c)
{
	int s;

	memset(c, 0, sizeof(rnd_sample_t));
	s = splhigh();
	pool_put(&rnd_mempool, c);
	splx(s);
}

/*
 * add a source to our list of sources
 */
void
rnd_attach_source(rs, name, type, flags)
	rndsource_element_t *rs;
	char *name;
	u_int32_t type;
	u_int32_t flags;
{
	u_int32_t ts;

	ts = rnd_timestamp();

	strcpy(rs->data.name, name);
	rs->data.last_time = ts;
	rs->data.last_delta = 0;
	rs->data.last_delta2 = 0;
	rs->data.total = 0;

	/*
	 * force network devices to not collect any entropy by
	 * default
	 */
	if (type == RND_TYPE_NET)
		flags |= (RND_FLAG_NO_COLLECT | RND_FLAG_NO_ESTIMATE);

	rs->data.type = type;
	rs->data.flags = flags;

	rs->data.state = rnd_sample_allocate(&rs->data);

	LIST_INSERT_HEAD(&rnd_sources, rs, list);

#ifdef RND_VERBOSE
	printf("%s: attached as an entropy source\n", rs->data.name);
#endif
}

/*
 * remove a source from our list of sources
 */
void
rnd_detach_source(rs)
	rndsource_element_t *rs;
{
	rnd_sample_t   *sample;
	rndsource_t    *source;
	int		s;

	s = splhigh();

	LIST_REMOVE(rs, list);

	source = &rs->data;

	if (source->state) {
		rnd_sample_free(source->state);
		source->state = NULL;
	}

	/*
	 * If there are samples queued up "remove" them from the sample queue
	 * by setting the source to the no-collect pseudosource.
	 */
	sample = SIMPLEQ_FIRST(&rnd_samples);
	while (sample != NULL) {
		if (sample->source == source)
			sample->source = &rnd_source_no_collect;

		sample = SIMPLEQ_NEXT(sample, next);
	}

	splx(s);
}

/*
 * Add a value to the entropy pool.  If rs is NULL no entropy estimation
 * will be performed, otherwise it should point to the source-specific
 * source structure.
 */
void
rnd_add_uint32(rs, val)
	rndsource_element_t *rs;
	u_int32_t val;
{
	rndsource_t    *rst;
	int		s;
	rnd_sample_t   *state;
	u_int32_t	ts;

	if (rs == NULL)
		return;

	rst = &rs->data;

	/*
	 * If we are not collecting any data at all, just return.
	 */
	if (rst->flags & RND_FLAG_NO_COLLECT)
		return;

	/*
	 * If the sample buffer is NULL, try to allocate one here.  If this
	 * fails, drop this sample.
	 */
	state = rst->state;
	if (state == NULL) {
		state = rnd_sample_allocate_isr(rst);
		if (state == NULL)
			return;
		rst->state = state;
	}

	/*
	 * Pick the timestamp.  If we are estimating entropy on this source,
	 * calculate differentials.
	 */
	ts = rnd_timestamp();
	if ((rst->flags & RND_FLAG_NO_ESTIMATE) == 0)
		state->entropy += rnd_estimate_entropy(rst, ts);

	state->ts[state->cursor] = ts;
	state->values[state->cursor] = val;
	state->cursor++;

	/*
	 * If the state arrays are not full, we're done.
	 */
	if (state->cursor < RND_SAMPLE_COUNT)
		return;

	/*
	 * State arrays are full.  Queue the state on the processing queue,
	 * and if the timeout isn't going, make it go.
	 */
	s = splhigh();
	SIMPLEQ_INSERT_HEAD(&rnd_samples, state, next);
	rst->state = NULL;

	if (rnd_timeout_pending == 0) {
		rnd_timeout_pending = 1;
		timeout(rnd_timeout, NULL, 1);
	}

	splx(s);

	/*
	 * To get here we have to have queued the state up, and therefore
	 * we need a new state buffer.  If we can, allocate one now.  Note
	 * that NULL pointers are not checked for here.
	 */
	rst->state = rnd_sample_allocate_isr(rst);
}

/*
 * timeout, run to process the events in the ring buffer.  Only one of these
 * can possibly be running at a time, and we are run at splsoftclock().
 */
static void
rnd_timeout(arg)
	void *arg;
{
	rnd_sample_t   *sample;
	rndsource_t    *source;

	rnd_timeout_pending = 0;

	sample = SIMPLEQ_FIRST(&rnd_samples);
	while (sample != NULL) {
		SIMPLEQ_REMOVE_HEAD(&rnd_samples, sample, next);

		source = sample->source;

		/*
		 * We repeat this check here, since it is possible the source
		 * was disabled before we were called, but after the entry
		 * was queued.
		 */
		if ((source->flags & RND_FLAG_NO_COLLECT)
		    == RND_FLAG_NO_COLLECT)
			goto loop;

		rndpool_add_data(&rnd_pool, sample->values,
				 RND_SAMPLE_COUNT * 4, 0);

		if ((source->flags & RND_FLAG_NO_ESTIMATE) == 0)
			rndpool_add_data(&rnd_pool, sample->ts,
					 RND_SAMPLE_COUNT * 4,
					 sample->entropy);
		else
			rndpool_add_data(&rnd_pool, sample->ts,
					 RND_SAMPLE_COUNT * 4, 0);

		source->total += sample->entropy;

	loop:
		rnd_sample_free(sample);
		sample = SIMPLEQ_FIRST(&rnd_samples);
	}

	/*
	 * wake up any potential readers waiting.
	 */
	rnd_wakeup_readers();
}

int
rnd_extract_data(p, len, flags)
	void *p;
	u_int32_t len;
	u_int32_t flags;
{
	int s;
	int retval;

	s = splsoftclock();

#if RND_USE_EXTRACT_TIME
	rndpool_add_uint32(&rnd_pool, rnd_timestamp(), 0);
#endif

	retval = rndpool_extract_data(&rnd_pool, p, len, flags);

	splx(s);

	return retval;
}
